sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -438,192 +423,169 @@ class GpuDynamicPartitionDataSingleWriter(
+                    val previousWriterStatus = currentWriterStatus
+                    if (previousWriterStatus != null) {
+                      releaseWriter(previousWriterStatus.outputWriter)
+                    }
+
+                    // create a new writer and update the writer in the status
+                    currentWriterStatus.outputWriter =
+                      newWriter(partPath, None, currentWriterStatus.fileCounter)
+                    currentWriterStatus.recordsInFile = 0
+                  }
+                  withResource(b.getTable()) {tab =>
+                    val bc = GpuColumnVector.from(tab, dataTypes)
+                    statsTrackers.foreach(_.newBatch(bc))
 
-                  currentWriterStatus.recordsInFile += b.getRowCount()
-                  currentWriterStatus.outputWriter.write(bc, statsTrackers)
-                  needNewWriter = true
-                }
-              })
+                    currentWriterStatus.recordsInFile += b.getRowCount()
+                    currentWriterStatus.outputWriter.write(bc, statsTrackers)
+                    needNewWriter = true
+                  }
+                })
+              }
             }
-          }
 
-        if (savedStatus.isDefined && savedStatus.get.tableCaches.nonEmpty) {
-          // convert caches seq to tables and close caches seq
-          val subTables = convertSpillBatchesToTables(savedStatus.get.tableCaches)
-          // clear the caches
-          savedStatus.get.tableCaches.clear()
-          // concat the caches and this `batch`
-          val concat = withResource(subTables) { _ =>
-            withResource(batch) { _ =>
-              // append `batch` to sub tables
-              subTables += GpuColumnVector.from(batch)
-            }
-            withResource(Table.concatenate(subTables: _*)) { concat =>
-              GpuColumnVector.from(concat, outDataTypes)
-            }
-          }
-          // write concat table
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            concat.numRows))
-          {
-            statsTrackers.foreach(_.newBatch(concat))
-            currentWriterStatus.recordsInFile += concat.numRows()
-            currentWriterStatus.outputWriter.write(concat, statsTrackers)
-          } else {
-            withResource(concat){concat =>  
-              writeBatch(concat)
-            }
-          }
-        } else {
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            batch.numRows))
-          {
-            statsTrackers.foreach(_.newBatch(batch))
-            currentWriterStatus.recordsInFile += batch.numRows()
-            currentWriterStatus.outputWriter.write(batch, statsTrackers)
-          } else {
-            withResource(batch){batch =>  
-              writeBatch(batch)
+          withResource(GpuColumnVector.from(table, outDataTypes)) { batch =>
```

Comment:
It seems to me this is created far too early, which is why we need to do the artificial refcount increments to work around the fact that this block is unconditionally going to close it.  Sometimes we want the `write` to close it directly, as in the original code.

Reply:
done